### PR TITLE
fix bugzilla 21995 Struct with size uint.max or greater causes ICE

### PR DESCRIPTION
**Pull request by WalterBright**
To its credit, the overflow at least was checked for!

### CODE REVIEW

File: compiler/src/dmd/aggregate.d [hunk 1]
```diff
@@ -829,7 +830,11 @@ public uint placeField(ref uint nextoffset, uint memsize, uint memalignsize,
     bool overflow;
     const sz = addu(memsize, actualAlignment, overflow);
     addu(ofs, sz, overflow);
-    if (overflow) assert(0);
+    if (overflow)
+    {
+        error(loc, "integer overflow adding %u + %u + %u when placing field in aggregate", memsize, actualAlignment, ofs);
```
Comment (ibuclaw) [inline compiler/src/dmd/aggregate.d:835]: While this error message correctly describes the bit of internal compiler logic that triggered it, it reads terribly when pairing it with the code that caused it.

The reason for why the code is rejected should be diagnosed to the user instead. That is, the size of the type exceeds more than half the address space or some max object size limit.

Comment (WalterBright) [reply compiler/src/dmd/aggregate.d]: ok

**Review by RazvanN7 — APPROVED** [review]
This should target stable.

File: compiler/src/dmd/aggregate.d [hunk 2]
```diff
@@ -829,7 +830,11 @@ public uint placeField(ref uint nextoffset, uint memsize, uint memalignsize,
     bool overflow;
     const sz = addu(memsize, actualAlignment, overflow);
     addu(ofs, sz, overflow);
-    if (overflow) assert(0);
+    if (overflow)
+    {
+        error(loc, "max object size %u exceeded from adding %u + %u + %u when placing field in aggregate", uint.max, memsize, actualAlignment, ofs);
```
Comment (JohanEngelen) [inline compiler/src/dmd/aggregate.d:835]: It'd be nice to explain the %u+%u+%u to the user (field size + alignment adjustment + field offset). Field name would be nice too...

Comment (WalterBright) [reply compiler/src/dmd/aggregate.d]: fixed

**Review by JohanEngelen — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**